lib/instrumentation/modules/pg.js

```diff
@@ -81,74 +82,43 @@ function patchClient (Client, klass, agent, enabled) {
           agent.logger.debug('unable to parse sql form pg module (type: %s)', typeof sql)
         }
 
+        const onQueryEnd = (_err) => {
+          agent.logger.debug('intercepted end of %s.prototype.%s %o', klass, name, { id: id })
+          span.end()
+        }
+
         if (typeof cb === 'function') {
-          args[index] = end
+          args[index] = agent._instrumentation.bindFunction((err, res) => {
+            onQueryEnd(err)
+            return cb(err, res)
+          })
           return orig.apply(this, arguments)
         } else {
-          cb = null
-          var query = orig.apply(this, arguments)
-
-          // The order of these if-statements matter!
-          //
-          // `query.then` is broken in pg <7 >=6.3.0, and since 6.x supports
-          // `query.on`, we'll try that first to ensure we don't fall through
-          // and use `query.then` by accident.
-          //
-          // In 7+, we must use `query.then`, and since `query.on` have been
-          // removed in 7.0.0, then it should work out.
-          //
-          // See this comment for details:
-          // https://github.com/brianc/node-postgres/commit/b5b49eb895727e01290e90d08292c0d61ab86322#commitcomment-23267714
-          if (typeof query.on === 'function') {
-            query.on('end', end)
-            query.on('error', end)
-          } else if (typeof query.then === 'function') {
-            query = query
-              .then(function (result) {
-                end()
-                return result
-              })
-              .catch(function (err) {
-                end()
-                throw err
-              })
+          var queryOrPromise = orig.apply(this, arguments)
+
+          // It is import to prefer `.on` to `.then` for pg <7 >=6.3.0, because
+          // `query.then` is broken in those versions. See
+          // https://github.com/brianc/node-postgres/commit/b5b49eb895727e01290e90d08292c0d61ab86322#r23267714
+          if (typeof queryOrPromise.on === 'function') {
+            queryOrPromise.on('end', onQueryEnd)
+            queryOrPromise.on('error', onQueryEnd)
+            if (queryOrPromise instanceof EventEmitter) {
+              agent._instrumentation.bindEmitter(queryOrPromise)
+            }
+          } else if (typeof queryOrPromise.then === 'function') {
+            queryOrPromise.then(
+              () => { onQueryEnd() },
+              onQueryEnd
+            )
```

Comment:
REVIEW NOTE: No need to pass back our modified promise because context tracking automatically ensures the `queryOrPromise` chain of handlers will run with the appropriate context.  Also no need to `throw err` in our reject/catch because we aren't returning this promise now.